_posts/2021-09-16-drain-cleaner.md

```diff
@@ -0,0 +1,195 @@
+controller is invoked after the schema validation phase and it calls every registered webhook, which implement some 
+logic. This can be useful to enforce custom policies or simply be notified about specific events.
+
+![admission controllers](/assets/images/posts/2021-09-16-admission-controllers.png)
+
+Some admission controllers are actually built-in (i.e. LimitRange, NamespaceLifecycle). ValidatingWebhooks are safe, as 
+they cannot change the incoming request and they always see the final version that would be persisted to etcd. 
+
+> Note: ValidatingWebhooks may be disabled in your Kubernetes cluster for security reasons, so this is the first thing to verify.
+
+In our case, we want the DrainCleaner to be notified of all Pod eviction requests. The application will then filter out 
+all Pods that are not part of our StatefulSet. One important note is that Kafka CR must have `maxUnavailable: 0` in 
+Kafka and Zookeeper [PodDisruptionBudgets](https://kubernetes.io/docs/concepts/workloads/Pods/disruptions/). This is 
+required to block Pod draining and let the combined work of DrainCleaner and ClusterOperator to handle this.
+
+This is how you can apply that configuration with a single patch command (no rolling update is needed):
+
+```sh
+kubectl patch kafka $CLUSTER_NAME --type json -p '[
+  {
+    "op": "add",
+    "path": "/spec/zookeeper/template",
+    "value": {
+      "PodDisruptionBudget": {
+        "maxUnavailable": 0
+      }
+    }
+  },
+  {
+    "op": "add",
+    "path": "/spec/kafka/template",
+    "value": {
+      "PodDisruptionBudget": {
+        "maxUnavailable": 0
+      }
+    }
+  },
+]'
+```
+
+## Validating webhook configuration
+
+To register the DrainCleaner's endpoint at the Kubernetes control-plane level, you need to create a 
+ValidatingWebhookConfiguration resource. This is a simple REST endpoint, exposed under `/drainer` URL path. When an 
+incoming API request matches one of the specified operations, groups, versions, resources, and scope for any of defined 
+rules, the request is sent to that endpoint.
+
+```yaml
+    rules:
+      - apiGroups:   [""]
+        apiVersions: ["v1"]
+        operations:  ["CREATE"]
+        resources:   ["Pods/eviction"]
+        scope:       "Namespaced"
+    clientConfig:
+      service:
+        namespace: "strimzi-drain-cleaner"
+        name: "strimzi-drain-cleaner"
+        path: /drainer
+        port: 443
+      caBundle: $CA_BUNDLE
+```
+
+Here I used a placeholder for the PEM encoded CA bundle, which is required to validate the webhook's certificate. You 
+can use a self-signed CA certificate, but the DrainCleaner's end-entity certificate, signed by that CA, must have a 
+common name (CN) that matches the application's service name.
+
+Let's see how that CA bundle looks like for a self-signed certificate deployed on a test namespace. CertManager or 
+OpenShift can help with this, by automatically generating and injecting the certificates (see GitHub repository).
```

Comment:
Is _(see GitHub repository)_ meant to be a link?

Reply:
You should ideally add a link here.

Reply:
ok